src/Corax/IndexWriter.cs

```diff
@@ -540,57 +560,68 @@ void DeleteIdFromExactTerm(long idToDelete, Slice fieldName, Span<byte> tmpBuffe
+                        return null;
+                    }
 
-                        //Fetch all the ids from the set into temporaryStorageForIds
-                        var itemsCount = ZigZagEncoding.Decode<int>(buffer, out var len);
-                        temporaryStorageForIds ??= new List<long>(itemsCount);
-                        temporaryStorageForIds.Clear();
+                    return containerId;
+                }
 
-                        long pos = len;
-                        var currentId = 0L;
+                if ((containerId & (long)TermIdMask.Small) != 0)
+                {
+                    var smallSetId = containerId & ~0b11;
+                    var buffer = Container.GetMutable(llt, smallSetId);
 
-                        while (pos < buffer.Length)
-                        {
-                            var delta = ZigZagEncoding.Decode<long>(buffer, out len, (int)pos);
-                            pos += len;
-                            currentId += delta;
-                            if (currentId == idToDelete)
-                                continue;
-                            temporaryStorageForIds.Add(currentId);
-                        }
+                    //Fetch all the ids from the set into temporaryStorageForIds
+                    var itemsCount = ZigZagEncoding.Decode<int>(buffer, out var len);
+                    temporaryStorageForIds ??= new List<long>(itemsCount);
+                    temporaryStorageForIds.Clear();
 
-                        // Due to encoding we have to encode new set again so we remove previous small set from container.
-                        Container.Delete(llt, _postingListContainerId, smallSetId);
-                        fieldTree.TryRemove(termValue, out var __); // term also disappears from the field tree
+                    long pos = len;
+                    var currentId = 0L;
 
-                        if (AddNewTerm(temporaryStorageForIds, tmpBuffer, out var termId))
-                            fieldTree.Add(termValue, termId);
-                    }
-                    else
+                    while (pos < buffer.Length)
                     {
-                        fieldTree.TryRemove(termValue, out var _);
+                        var delta = ZigZagEncoding.Decode<long>(buffer, out len, (int)pos);
+                        pos += len;
+                        currentId += delta;
+                        if (currentId == idToDelete)
+                            continue;
+                        temporaryStorageForIds.Add(currentId);
                     }
+
+                    // Due to encoding we have to encode new set again so we remove previous small set from container.
+                    Container.Delete(llt, _postingListContainerId, smallSetId);
+
+                    return AddNewTerm(temporaryStorageForIds, tmpBuffer, out var termId) ? termId : null;
                 }
+
+                return null;
```

Comment:
There is an issue here, we need to also remove the the element if it is the only value in the tree (not a container or set).

Reply:
Yes, that is being done in the caller, when you return `null`, we remove the previous value.